posthog/clickhouse/cluster.py

```diff
@@ -172,11 +171,16 @@ def map_all_hosts_in_shards(
         Wait for all to return before returning upon ``.values()``
         """
 
-        return reduce(
-            lambda x, y: x.merge(y),
-            [self.map_all_hosts_in_shard(shard, fn, concurrency) for shard, fn in shard_fns],
-            FuturesMap(),
-        )
+        shard_host_fn = {}
+        for shard, fn in shard_fns.items():
+            for host in self.__hosts:
+                if host.shard_num == shard:
+                    shard_host_fn[host] = fn
```

Comment:
It'd probably be good to error here if a shard is provided that doesn't actually exist

I was planning on moving things and replacing `__hosts` with a `__shard_hosts` map + `__extra_hosts` list (which maybe we get rid of with the new coordinator-aware cluster) - that could help here since an invalid shard would `KeyError` when trying to find it in the map, but that would also be a bigger refactoring bite than maybe you want to take here